packages/test/test-end-to-end-tests/src/test/gc/gcStateResetInSummaries.spec.ts

```diff
@@ -159,42 +217,163 @@ describeNoCompat("GC state reset in summaries", (getTestObjectProvider) => {
+            undefined /* unreferencedDataStoreIds */,
+            true /* shouldGCDataBeHandle */,
+        );
+
+        // Load a new summarizer from the last summary with GC disabled.
         summarizer2.close();
-        const summarizer3 = await createSummarizer(provider, mainContainer, summaryVersion);
-        // Validate that GC runs and the summary is regenerated because GC was disabled in the previous summary and
-        // is now enabled. Also, the unreferenced data stores should be marked as such.
+        const summarizer3 = await createSummarizer(provider, mainContainer, summaryVersion, { disableGC: true });
+        // Validate that GC does not run and the summary is regenerated.
         await summarizeAndValidateGCState(
             summarizer3,
+            false /* shouldGCRun */,
+            true /* shouldRegenerateSummary */,
+        );
+    });
+
+    /**
+     * This test validates that the GC state is regenerated if needed when state is refreshed from a snapshot. If GC is
+     * enabled and state is refreshed from a snapshot that had GC disabled, the GC state needs to be reset and a full
+     * tree summary should happen.
+     */
+    it("regenerates GC state on refreshing from snapshot that has GC disabled", async () => {
+        // Create a document with GC allowed. It has to be allowed on creation because this setting cannot be changed
+        // throughout the lifetime of the document.
+        mainContainer = await createContainer(true /* gcAllowed */);
+        const mainDataStore = await requestFluidObject<ITestDataObject>(mainContainer, "default");
+        await waitForContainerConnection(mainContainer);
+
+        // Create a data store and mark it as unreferenced by storing and the removing its handle in a referenced DDS.
+        const newDataStore = await requestFluidObject<ITestDataObject>(
+            await mainDataStore._context.containerRuntime.createDataStore(TestDataObjectType), "");
+        mainDataStore._root.set("newDataStore", newDataStore.handle);
+        mainDataStore._root.delete("newDataStore");
+
+        // Create a summarizer with GC enabled and summarize.
+        const { summarizer: summarizerGCEnabled } = await createSummarizerAndContainer();
+
+        // Summarize with GC enabled and validate that GC ran.
+        let summaryVersion = await summarizeAndValidateGCState(
+            summarizerGCEnabled,
+            true /* shouldGCRun */,
+            true /* shouldRegenerateSummary */,
+            [newDataStore._context.id],
+        );
+
+        // Create a summarizer with GC disabled and another one with GC enabled from the above summary where GC was
+        // enabled.
+        const { container: containerGCDisabled, summarizer: summarizerGCDisabled } =
+            await createSummarizerAndContainer(summaryVersion, true /* gcDisabled */);
+        const { container: containerGCEnabled2, summarizer: summarizerGCEnabled2 } =
```

Comment:
What if you created this summarizer after running 
```
        summaryVersion = await summarizeAndValidateGCState(
            summarizerGCDisabled,
            false /* shouldGCRun */,
            true /* shouldRegenerateSummary */,
        );
```
I don't think you would need to reconnect the summarizer. It would seem that containerGCEnabled2 is the main summarizer.